cargo/src/audit.rs

```diff
@@ -73,24 +78,135 @@ pub struct AuditOpts {
     pub color: clap::ColorChoice,
 }
 
+#[derive(Debug)]
+pub struct AuditFailed {
+    error: String,
+    boxy: Box<dyn Error>,
+}
+
+impl Display for AuditFailed {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let msg = format!("{}. Got {}", self.error, self.boxy);
+        write!(f, "{}", msg)
+    }
+}
+
+impl AuditOpts {
+    #[allow(dead_code)]
+    pub fn new(
+        self,
+        lockfiles: Vec<PathBuf>,
+        outdir: Option<PathBuf>,
+        color: clap::ColorChoice,
+    ) -> Self {
+        Self {
+            lockfile: lockfiles,
+            outdir,
+            color,
+        }
+    }
+
+    pub fn generate_opts(self, p: &Path) -> io::Result<Vec<Opts>> {
+        make_opts(p)
+    }
+}
+
+impl Error for AuditFailed {}
+
 pub trait Audit {
-    fn run_audit(self, opts: &AuditOpts) -> io::Result<()>;
-    fn process_lockfiles(self) -> io::Result<()>;
+    // Run audit sets the workdir before running other stuff
+    fn run_audit(self, opts: &Opts) -> Result<(), AuditFailed>;
+    fn audit_vendored_tar(self, opts: &Opts, workdir: &Path) -> io::Result<()>;
```

Comment:
You don't need to have opts for this future self, just extract and get the Cargo.lock. You also need to check if there are multiple vendored tarballs. So in this case, use the `glob` crate and use `/**vendor**.tar.**/` and check their lockfiles.

Reply:
Ahhh true. That's a good point.